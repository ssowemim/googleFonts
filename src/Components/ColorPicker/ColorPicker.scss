.ColorPicker {
  position: relative;
  .ColorPicker__Popover {
    position: absolute;
    z-index: 2;
    right: 0px; 
  }
  
  .ColorPicker__Swatch {
    padding: 5px;
    background: #fff;
    border-radius: 4px;
    height: 36px;
    width: 45px;
    border: 1px solid rgba(0,0,0,.1);
    display: inline-block;
    cursor: pointer;
    &.is-focused { 
      border-color: #ffe000;
      box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075), 0 0 0 3px rgba(255, 224, 0, 0.2);
    }
  }
  
  .ColorPicker__Color {
    width: 100%;
    height: 100%;
    border-radius: 2px;
  }
}