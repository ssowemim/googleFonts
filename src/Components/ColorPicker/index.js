import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { SketchPicker } from 'react-color';
import enhanceWithClickOutside from 'react-click-outside';
import './ColorPicker.css';

class ColorPicker extends Component {
    constructor(props) {
        super(props);
        this.state = {
            color: this.props.color,
            displayColorPicker: false,
        }
        this.handleOnClick = this.handleOnClick.bind(this);
        this.handleOnChange = this.handleOnChange.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({ color: nextProps.color })  
    }

    handleOnClick() {
        this.setState({ displayColorPicker: !this.state.displayColorPicker })
    }

    handleOnChange(color) {
        this.setState({ color: color.rgb }, () => {
            if (this.props.isBackgroundColorPicker) {
                document.body.style.backgroundColor = `rgba(${this.state.color.r}, ${this.state.color.g}, ${this.state.color.b}, ${this.state.color.a})`;
            }

            else {
                this.props.handleChangeColor(color)
            }
        })
    }

    handleClickOutside() {
        this.setState({ displayColorPicker: false })
    }

    __getStyleObject(isBackgroundColorPicker) {
        const colorObject = {
            color: `rgba(${this.state.color.r}, ${this.state.color.g}, ${this.state.color.b}, ${this.state.color.a})`,
        };

        const backGroundObject = {
            backgroundColor: `rgba(${this.state.color.r}, ${this.state.color.g}, ${this.state.color.b}, ${this.state.color.a})`,
        };

        return isBackgroundColorPicker ? backGroundObject : colorObject

    }

    render() {
        return (
            <div className="ColorPicker">
                <div className={`${this.state.displayColorPicker ? "is-focused " : ""}ColorPicker__Swatch`} onClick={this.handleOnClick}>
                    <div className="ColorPicker__Color" style={this.__getStyleObject(this.props.isBackgroundColorPicker)}>
                        {!this.props.isBackgroundColorPicker && "A"}
                    </div>
                </div>
                {this.state.displayColorPicker ?
                    <div className="ColorPicker__Popover">
                        <SketchPicker color={this.state.color} onChange={this.handleOnChange} />
                    </div> :
                    null}

            </div>
        );
    }
}

ColorPicker.propTypes = {
    isBackgroundColorPicker: PropTypes.bool,
    handleChangeColor: PropTypes.func
};

export default enhanceWithClickOutside(ColorPicker);