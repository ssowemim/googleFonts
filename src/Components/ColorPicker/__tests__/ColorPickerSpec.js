import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { SketchPicker } from 'react-color';
import ColorPicker from './../index';

describe('ColorPicker', () => {
    let colorPickerProps;

    beforeEach(()=>{
        colorPickerProps = {
            r: '0',
            g: '0',
            b: '0',
            a: '1',
        }
    });

    it('renders correctly with color objct', () => {
        let tree = renderer.create(
            <ColorPicker color={colorPickerProps} />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with bgColor objct', () => {
        colorPickerProps.isBackgroundColorPicker = true;
        let tree = renderer.create(
            <ColorPicker color={colorPickerProps} isBackgroundColorPicker/>
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });
    

    it('handleOnClick should toggle displayColorPicker state', () =>{
        let wrapper = shallow(<ColorPicker color={colorPickerProps} />).first().shallow();
        let elem = wrapper.find('.ColorPicker__Swatch');
        elem.simulate("click");
        expect(wrapper.instance().state.displayColorPicker).toBeTruthy()
        elem.simulate("click");
        expect(wrapper.instance().state.displayColorPicker).toBeFalsy()
    });

    it('handleOnChange should set body bg if isBackgroundColorPicker is true', ()=>{
        let wrapper = shallow(<ColorPicker color={colorPickerProps} isBackgroundColorPicker />).first().shallow();
        wrapper.instance().handleOnChange({ rgb:{r: '0', g: '0', b: '0', a: '.3' }});
        expect(wrapper.instance().state.color).toEqual({"a": ".3", "b": "0", "g": "0", "r": "0"});
        expect(document.body.style.backgroundColor).toEqual("rgba(0, 0, 0, 0.3)")
        document.body.style = null
    })

    it('handleOnChange should call call back if isBackgroundColorPicker is false', ()=>{
        let handleChangeColor = jest.fn();
        let element = shallow(<ColorPicker color={colorPickerProps} isBackgroundColorPicker={false} handleChangeColor={handleChangeColor} />).first().shallow();
        element.update();
        element.instance().handleOnChange({ rgb:{r: '0', g: '0', b: '0', a: '.3' }});
        expect(document.body.style.backgroundColor).toEqual("");
        expect(handleChangeColor).toHaveBeenCalled()
    })


    it('handleClickOutside should set displayColorPicker state to false', ()=>{
        const spy = jest.spyOn(ColorPicker.prototype, 'handleClickOutside');
        let wrapper = shallow(<ColorPicker color={colorPickerProps} />);
        document.body.click();
        expect(wrapper.instance().state.displayColorPicker).toBeFalsy()
        expect(spy).toHaveBeenCalled()
    });


});
