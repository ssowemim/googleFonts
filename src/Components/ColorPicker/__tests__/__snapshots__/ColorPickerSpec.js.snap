// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker renders correctly with bgColor objct 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__Swatch"
    onClick={[Function]}
  >
    <div
      className="ColorPicker__Color"
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 1)",
        }
      }
    />
  </div>
</div>
`;

exports[`ColorPicker renders correctly with color objct 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__Swatch"
    onClick={[Function]}
  >
    <div
      className="ColorPicker__Color"
      style={
        Object {
          "color": "rgba(0, 0, 0, 1)",
        }
      }
    >
      A
    </div>
  </div>
</div>
`;
